diff --git a/src/runtime/crash_nonunix_test.go b/src/runtime/crash_nonunix_test.go
index 2ce995c..3689ac4 100644
--- a/src/runtime/crash_nonunix_test.go
+++ b/src/runtime/crash_nonunix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build windows plan9 nacl
+// +build windows plan9 nacl atman
 
 package runtime_test
 
diff --git a/src/runtime/lock_sema.go b/src/runtime/lock_sema.go
index 0fa0481..42834cf 100644
--- a/src/runtime/lock_sema.go
+++ b/src/runtime/lock_sema.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin nacl netbsd openbsd plan9 solaris windows
+// +build darwin nacl netbsd openbsd plan9 solaris windows atman
 
 package runtime
 
diff --git a/src/runtime/write_err.go b/src/runtime/write_err.go
index 6b1467b..5678ced 100644
--- a/src/runtime/write_err.go
+++ b/src/runtime/write_err.go
@@ -3,6 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build !android
+// +build !atman
 
 package runtime
 
